### PR TITLE
Fix SFTDataset sample generation

### PR DESCRIPTION
**Pull request by alealv**
Fixes https://github.com/Lightning-AI/litgpt/issues/1567.

### CODE REVIEW

File: litgpt/data/base.py [hunk 1]
```diff
@@ -79,11 +79,9 @@ def __getitem__(self, idx: int) -> Dict[str, Tensor]:
         if self.transform is not None:
             example = self.transform(example)
         prompt = self.prompt_style.apply(prompt=example["instruction"], **example)
-        prompt_and_response = prompt + example["output"]
         encoded_prompt = self.tokenizer.encode(prompt, max_length=self.max_seq_length)
-        encoded_prompt_and_response = self.tokenizer.encode(
-            prompt_and_response, eos=True, max_length=self.max_seq_length
-        )
+        encoded_response = self.tokenizer.encode(example["output"], eos=True, max_length=self.max_seq_length)
+        encoded_prompt_and_response = encoded_prompt + encoded_response[1:] # We remove bos from response
```
Comment (rasbt) [inline litgpt/data/base.py:84]: Thanks for the PR! Don't we also need a truncation at the end now? I.e.,

```suggestion
        encoded_prompt_and_response = encoded_prompt + encoded_response[1:self.max_seq_length]  # We remove bos from response
```

Comment (rasbt) [inline litgpt/data/base.py:84]: Another thought: Shouldn't we check that the BOS sequence was added in the first place (which may not always be true)?

E.g., there's `tokenizer.use_bos`

**Comment by rasbt** [comment]
Also tagging @Andrei-Aksionov  for insights here, who has done a lot of tokenizer work in LitGPT

File: litgpt/data/base.py [hunk 2]
```diff
@@ -79,11 +79,9 @@ def __getitem__(self, idx: int) -> Dict[str, Tensor]:
         if self.transform is not None:
             example = self.transform(example)
         prompt = self.prompt_style.apply(prompt=example["instruction"], **example)
-        prompt_and_response = prompt + example["output"]
         encoded_prompt = self.tokenizer.encode(prompt, max_length=self.max_seq_length)
-        encoded_prompt_and_response = self.tokenizer.encode(
-            prompt_and_response, eos=True, max_length=self.max_seq_length
-        )
+        encoded_response = self.tokenizer.encode(example["output"], eos=True, max_length=self.max_seq_length)
```
Comment (awaelchli) [inline litgpt/data/base.py:83]: By default, encode only adds a bos if the config specifies it should do so. Therefore, after these changes you need to set 
```suggestion
        encoded_response = self.tokenizer.encode(example["output"], eos=True, bos=False, max_length=self.max_seq_length)
```

here and remove this logic:

```
encoded_response[1:]
```

**Comment by Andrei-Aksionov** [comment]
Hello @alealv 
Nice catch!
Just update the test and I think the PR is ready.

**Review by Andrei-Aksionov — APPROVED** [review]

**Pull request merged**